core/src/Volume.cpp

```diff
@@ -202,17 +202,28 @@ Reslice Volume::reslice(
 
 cv::Mat Volume::load_slice_(int index) const
 {
+    std::cout << "Requested to load slice " << index << std::endl;
     auto slicePath = getSlicePath(index);
     return cv::imread(slicePath.string(), -1);
 }
 
 cv::Mat Volume::cache_slice_(int index) const
 {
-    if (cache_->contains(index)) {
-        return cache_->get(index);
+    {
+        std::shared_lock<std::shared_mutex> lock(cache_mutex_);
+        if (cache_->contains(index)) {
+            return cache_->get(index);
+        }
```

Comment:
Performance should be okay, since the chache->get itself is quite quick. Changed it to unique lock for the whole function and removed the unnecessary get call.